File: Xamarin/Common/Presenter/PropertyFinderPresenter.cs
using System;
using System.Collections.Generic;
using PropertyFinder.Model;

namespace PropertyFinder.Presenter
{
  /// <summary>
  /// A presenter for the front-page of this application This presenter allows the
  /// user to search by a text string or their current location.
  /// </summary>
  public class PropertyFinderPresenter
  {
    /// <summary>
    /// The interface this presenter requires from the assocaited view.
    /// </summary>
    public interface View
    {
      /// <summary>
      /// Sets the text displayed in the search field.
      /// </summary>
      string SearchText { set; }

      /// <summary>
      /// Supplies a message to the user, typically to indicate an error or problem.
      /// </summary>
      void SetMessage(string message);

      /// <summary>
      /// Displays a list of suggested locations when the user supplies a plain-text search.
      /// </summary>
      void DisplaySuggestedLocations(List<Location> locations);

      /// <summary>
      /// Displays a list of recently performed searches
      /// </summary>
      void DisplayRecentSearches(List<RecentSearch> recentSearches);

      /// <summary>
      /// Sets whether to display a loading indicator
      /// </summary>
      bool IsLoading { set; }

      event EventHandler SearchButtonClicked;

      event EventHandler<SearchTextChangedEventArgs> SearchTextChanged;

      event EventHandler MyLocationButtonClicked;

      event EventHandler FavouritesClicked;

      event EventHandler<LocationSelectedEventArgs> LocationSelected;

      event EventHandler<RecentSearchSelectedEventArgs> RecentSearchSelected;
    }

    private View _view;

    private PropertyDataSource _propertyDataSource;

    private INavigationService _navigationService;

    private PropertyFinderPersistentState _state;

    private IGeoLocationService _geolocationService;

    private SearchItemBase _searchItem = new PlainTextSearchItem("");

    public PropertyFinderPresenter(PropertyFinderPersistentState state,
      PropertyDataSource dataSource, INavigationService navigationService, IGeoLocationService geolocationService)
    {
      _propertyDataSource = dataSource;
      _navigationService = navigationService;
      _state = state;
      _geolocationService = geolocationService;
    }

    public void SetView(View view)
    {
      _view = view;
      _view.SearchButtonClicked += View_SearchButtonClicked;
      _view.LocationSelected += View_LocationSelected;
      _view.FavouritesClicked += View_FavouritesClicked;
      _view.MyLocationButtonClicked += View_MyLocationButtonClicked;
      _view.SearchTextChanged += View_SearchTextChanged;
      _view.RecentSearchSelected += View_RecentSearchSelected;

      _view.DisplayRecentSearches(_state.RecentSearches);
    }

    private void View_RecentSearchSelected (object sender, RecentSearchSelectedEventArgs e)
    {
      _searchItem = e.RecentSearch.Search;
      _view.SearchText = e.RecentSearch.Search.DisplayText;
      SearchForProperties();
    }
    
    private void View_SearchTextChanged(object sender, SearchTextChangedEventArgs e)
    {
      if (e.Text != _searchItem.DisplayText)
      {
        _searchItem = new PlainTextSearchItem(e.Text);
      }
    }

    private void View_MyLocationButtonClicked(object sender, EventArgs e)
    {
      _view.IsLoading = true;
      _geolocationService.GetLocation(location =>
        {
          _searchItem = new GeoLocationSearchItem(location);
          _view.SearchText = _searchItem.DisplayText;
          _view.IsLoading = false;

          SearchForProperties();
        });
    }

    private void View_FavouritesClicked(object sender, EventArgs e)
    {
      var presenter = new FavouritesPresenter(_navigationService, _state);
      _navigationService.PushPresenter(presenter);
    }

    private void View_LocationSelected(object sender, LocationSelectedEventArgs e)
    {
      _view.SearchText = e.Location.DisplayName;
      _view.DisplaySuggestedLocations(null);
      _searchItem = new PlainTextSearchItem(e.Location.Name, e.Location.DisplayName);
      SearchForProperties();
    }

    private void View_SearchButtonClicked(object sender, EventArgs e)
    {
      SearchForProperties();
    }

    private void SearchForProperties()
    {
      _view.IsLoading = true;

      _view.SetMessage(null);

      _searchItem.FindProperties(_propertyDataSource, 1, response =>
      {
        if (response is PropertyListingsResult)
        {
          var propertiesResponse = (PropertyListingsResult)response;
          if (propertiesResponse.Data.Count == 0)
          {
            _view.SetMessage("There were no properties found for the given location.");
          }
          else
          {
            var listingsResponse = (PropertyListingsResult)response;
            _state.AddSearchToRecent(new RecentSearch(_searchItem, listingsResponse.TotalResult));
            _view.DisplayRecentSearches(_state.RecentSearches);
            var presenter = new SearchResultsPresenter(_navigationService, _state, listingsResponse,
                                                        _searchItem, _propertyDataSource);
            _navigationService.PushPresenter(presenter);
          }
        }
        else if (response is PropertyLocationsResult)
        {
          _view.DisplayRecentSearches(null);
          _view.DisplaySuggestedLocations(((PropertyLocationsResult)response).Data);
        }
        else
        {
          _view.SetMessage("The location given was not recognised.");
        }

        _view.IsLoading = false;
      });
    }
  }
}

File: native/android/PropertyCross/src/com/propertycross/android/presenter/PropertyFinderPresenter.java
package com.propertycross.android.presenter;

import java.util.List;

import com.propertycross.android.events.Callback;
import com.propertycross.android.events.LocationSelectedEvent;
import com.propertycross.android.events.LocationSelectedEventArgs;
import com.propertycross.android.events.RecentSearchSelectedEvent;
import com.propertycross.android.events.RecentSearchSelectedEventArgs;
import com.propertycross.android.events.SearchTextChangedEvent;
import com.propertycross.android.events.SearchTextChangedEventArgs;
import com.propertycross.android.events.UIEvent;
import com.propertycross.android.model.Location;
import com.propertycross.android.model.PropertyDataSource;
import com.propertycross.android.model.PropertyDataSourceResult;
import com.propertycross.android.model.PropertyListingsResult;
import com.propertycross.android.model.PropertyLocationsResult;
import com.propertycross.android.presenter.searchitems.GeoLocationSearchItem;
import com.propertycross.android.presenter.searchitems.PlainTextSearchItem;
import com.propertycross.android.presenter.searchitems.SearchItem;

public class PropertyFinderPresenter {

	public interface View {
		void setSearchText(String searchText);
		void setMessage(String message);
		void displaySuggestedLocations(List<Location> locations);
		void displayRecentSearches(List<RecentSearch> recentSearches);
		void setIsLoading(boolean isLoading);
		void searchButtonClicked(Callback<UIEvent> callback);
		void searchTextChanged(Callback<SearchTextChangedEvent> callback);
		void myLocationButtonClicked(Callback<UIEvent> callback);
		void favouritesClicked(Callback<UIEvent> callback);
		void locationSelected(Callback<LocationSelectedEvent> callback);
		void recentSearchSelected(Callback<RecentSearchSelectedEvent> callback);
	}

	private View view;
	private PropertyDataSource propertyDataSource;
	private INavigationService navigationService;
	private PropertyFinderPersistentState state;
	private IGeoLocationService geolocationService;
	private SearchItem searchItem = new PlainTextSearchItem();

	public PropertyFinderPresenter(PropertyFinderPersistentState state, PropertyDataSource dataSource,
			INavigationService navigationService, IGeoLocationService geolocationService) {
		this.propertyDataSource = dataSource;
		this.navigationService = navigationService;
		this.state = state;
		this.geolocationService = geolocationService;
	}

	public void setView(View v) {
		view = v;

		view.searchButtonClicked(new Callback<UIEvent>() {
			
			public void complete(UIEvent event) {
				searchForProperties();
			}
		});
		
		view.searchTextChanged(new Callback<SearchTextChangedEvent>() {
			
			public void complete(SearchTextChangedEvent event) {
				SearchTextChangedEventArgs args = (SearchTextChangedEventArgs) event.getArguments();
				if (args.getText() != searchItem.getDisplayText())
					searchItem = new PlainTextSearchItem(args.getText());
			}
		});
		
		view.myLocationButtonClicked(new Callback<UIEvent>() {
			
			public void complete(UIEvent event) {
				view.setIsLoading(true);
				geolocationService.getLocation(new Callback<GeoLocation> () {

					@Override
					public void complete(GeoLocation location) {
						view.setIsLoading(false);
						if (location == null) {
							view.setMessage("Unable to detect current location. Please ensure location is turned on in your phone settings and try again.");
						}
						else {
							searchItem = new GeoLocationSearchItem(location);
							view.setSearchText(searchItem.getDisplayText());
							searchForProperties();
						}
					}
					
				});
			}
		});
		
		view.favouritesClicked(new Callback<UIEvent>() {
			
			public void complete(UIEvent event) {
				FavouritesPresenter presenter = new FavouritesPresenter(navigationService, state);
				navigationService.pushPresenter(presenter);
			}
		});
		
		view.locationSelected(new Callback<LocationSelectedEvent>() {
			
			public void complete(LocationSelectedEvent event) {
				Location location = ((LocationSelectedEventArgs) event.getArguments()).getLocation();
				view.setSearchText(location.getDisplayName());
				view.displaySuggestedLocations(null);
				searchItem = new PlainTextSearchItem(location.getName(),location.getDisplayName());
				PropertyFinderPresenter.this.searchForProperties();
			}
		});
		
		view.recentSearchSelected(new Callback<RecentSearchSelectedEvent>() {
			
			public void complete(RecentSearchSelectedEvent event) {
				RecentSearch search = ((RecentSearchSelectedEventArgs) event.getArguments()).getRecentSearch();
				searchItem = search.getSearch();
				view.setSearchText(searchItem.getDisplayText());
				PropertyFinderPresenter.this.searchForProperties();
			}
		});
		
		view.displayRecentSearches(state.getRecentSearches().size() > 0 ? state.getRecentSearches() : null);
	}

	private void searchForProperties() {
		view.setIsLoading(true);
		searchItem.findProperties(propertyDataSource, 1,
				new Callback<PropertyDataSourceResult>() {

					@Override
					public void complete(PropertyDataSourceResult result) {
						
						if(result instanceof PropertyListingsResult) {
							PropertyListingsResult propertiesResult = (PropertyListingsResult) result;
							
							if(propertiesResult.getData().size() == 0) {
								view.setMessage("There were no properties found for the given location.");
							}
							else {
								state.addSearchToRecent(new RecentSearch(searchItem, propertiesResult.getTotalResult()));
								view.displayRecentSearches(state.getRecentSearches());
								SearchResultsPresenter presenter = 
										new SearchResultsPresenter(navigationService, state, propertiesResult, searchItem, propertyDataSource);
								navigationService.pushPresenter(presenter);
							}
						}
						else if (result instanceof PropertyLocationsResult) {
							view.displayRecentSearches(null);
							view.displaySuggestedLocations(((PropertyLocationsResult) result).getData());
						}
						else {
							view.setMessage("The location given was not recognised");
						}
						
						view.setIsLoading(false);						
					}
			
		},
				new Callback<Exception>() {

					@Override
					public void complete(Exception e) {
						view.setMessage("An error occurred while searching. Please check your network connection and try again.");
						view.setIsLoading(false);
					}
		});		
	}
}
